### ISSUE TITLE
Problems with wrapper classes

### ISSUE THREAD

**Issue by warsaw** [python/mypy#5523]
Let's say I have a `Person` class with a bunch of attributes; it comes from a library I don't control but for which type information exists.  Now I have a `PersonWrapper` class which provides for defaults so that accessing missing attributes never raises an exception (think, use of instances in f-strings).

```
class PersonWrapper:
    def __init__(self, person: Person, default: Any = '') -> None:
        self.person = person
        self.default = default

    def __getattr__(self, attribute: str) -> Any:
        return getattr(self.person, attribute, self.default)
```

In general, any place I want to use a `Person` I might also want to use a `PersonWrapper`, so I have variables that can be assigned either a `Person` or a `PersonWrapper`.  mypy seems not to be able to handle this, and I don't really know of any way to tell the type checker that it should treat `Person` and `PersonWrapper` as equivalent.  For example, if I have this code:

```
def print_person(original_person: Person) -> None:
    person = PersonWrapper(original_person)
    # This attribute might be missing from the original person object.
    print(f'  Username: {person.username}')
    person = original_person
    # Only print this if the attribute exists.
    phone = getattr(person, 'phoneNumber', None)
    if phone:
        print(f'    Phone: {phone}')
```

mypy complains about the assignment of `original_person` to `person`:

`error: Incompatible types in assignment (expression has type "Person", variable has type "PersonWrapper")`

Because `Person` has some dynamic attribute access as well, I can't change the original assignment to:

```
    person: Person = cast(Person, PersonWrapper(original_person))
```

because then I get errors such as:

`error: "Person" has no attribute "office"`

The way I fix this is by changing the second assignment to use `cast()`:

```
    person = cast(PersonWrapper, original_person)
```

This feels wrong, but maybe it's okay.  What I think i'd like to do is to create an equivalence between `PersonWrapper` and `Person`, but then I'd probably still have to handle the dynamic parts of `Person` in some way.  FWIW I didn't have much luck using `Union[Person, PersonWrapper]` either.

Is there a better way to handle wrapper classes?

**Comment by ilevkivskyi** [comment]
This is a quite common feature request, see e.g. https://github.com/python/mypy/issues/2087#issuecomment-392904721 (I prefer to call these proxy classes). I am not sure we need to keep this open, it is almost a duplicate of #2087 in the sense that implementing any sort of `@allow_attributes_of(T)` will fix both issues.

Btw, there is a chance this will be fixed/implemented in October.

**Comment by ilevkivskyi** [comment]
Btw, there is an aspect of proxy types not discussed in #2087 (where the focus is mostly on attribute access): should we accept the proxy where the original is expected? For example:
```python
class A: ...

@allow_attributes_of(A)
class B:
    def __getattr__(self, attr: str) -> Any:
        return getattr(A(), attr)

x: A = B()
```
The above will naturally work (when we implement attribute access) if `A` is a protocol, but the question is whether this should be allowed if `A` is _not_ a protocol? As shown in OP this may be a common situation. On one hand it will work at runtime in all reasonable situations, but on another hand, this can confuse tools like mypyc that take nominal types seriously.

@JukkaL what do you think?

**Comment by JukkaL** [comment]
I think that we should make proxys valid when the original type is expected. Otherwise `@allow_attributes` may not be very useful. There are precedents that are similar in spirit -- `str` is treated as compatible with `unicode` in Python 2, even though there is no subclass relationship.

**Comment by ilevkivskyi** [comment]
Should we just re-use `_promote` for this? Or do we need another plugin hook? (I think we can implement `@allow_attributes` using a plugin).

**Comment by JukkaL** [comment]
`_promote` might be the best alternative. The implementation may need some refinement, since it's only been used for a few things in the stdlib and some edge cases might not work.

**Comment by ilevkivskyi** [comment]
> The implementation may need some refinement, since it's only been used for a few things in the stdlib and some edge cases might not work.

Yes, for example generic proxies will definitely need something like `expand_type_by_instance` applied to `_promote`. But I think it should be straightforward to implement (we need to be careful to generate correct fine grained dependencies however, i.e., on both proxy and target).

**Comment by sirosen** [comment]
Is there a current solution or workaround for this problem? I tried writing a plugin for my case, but I found that in `get_attribute_hook`, I couldn't get the name of the attribute being accessed (maybe I missed it, but it looks like `AttributeContext` doesn't include the name).

There are cases of proxy-like objects which apply well-defined type transformations on the underlying types. e.g.
```python
P = ParamSpec("P")
T = TypeVar("T")

def listify(func: Callable[P, T] -> Callable(P, List[T]):
    def inner(*args: P.args, **kwargs: P.kwargs) -> List[T]:
        return [func(*args, **kwargs)]
    return inner


class ListifyingCallTransform:
    def __init__(self, wrapped):
        self._wrapped = wrapped

    def __getattr__(self, key: str) -> Any:
        orig = getattr(self._wrapped, key)
        if inspect.ismethod(orig):
            return listify(orig)
        return orig
```

Notably, that makes this usage undesirable:

```python
class A:
    def foo() -> int:
        return 1

x = cast(A, ListifyingCallTransform(A()))
```

Although it means that `x.foo()` would be typed correctly as taking no arguments, it would be understood to be of type `int`, rather than `List[int]`.

Is there a way to get correct types for a class by ListifyingCallTransform by writing a plugin?

**Comment by ocmob** [comment]
Are there any plans to implement this?

**Comment by Waghabond** [comment]
Here's a kinda hacky work around. 
```python
import datetime
from typing import Any, cast, Generic, TypeVar

Proxied = TypeVar("Proxied")

class TimestampLocaliserProxy(Generic[Proxied]):
    """Proxy class that wraps any object and localises any datetime attributes of that object to the specified timezone
    as it returns them"""

    def __init__(self, obj: Proxied, tz_info: zoneinfo.ZoneInfo):
        self._tz_info = tz_info
        self._obj = obj

    def __getattr__(self, name: str) -> Any:
        if not self._obj:
            raise AttributeError("Proxy has not been provided an object to wrap.")
        attr = getattr(self._obj, name)
        if isinstance(attr, datetime.datetime):
            return attr.astimezone(self._tz_info)
        return attr

    def unwrap(self) -> Optional[Proxied]:
        """returns the wrapped object"""
        return self._obj


def wrap_with_timestamp_localiser(obj: Proxied, tz_info: zoneinfo.ZoneInfo) -> Proxied:
    """Wrap's the given object and returns TimestampLocaliserProxy pretending to be the proxied object

    Args:
        obj (Proxied): object to wrap.
        tz_info (zoneinfo.ZoneInfo): Timezone

    Returns:
        Proxied: this TimezoneLocaliser pretending to be the same type as the proxied object.
    """
    return cast(Proxied, TimestampLocaliserProxy(obj, tz_info))
```

Here's an example.
```python
class MyClass:
    def __init__(self, non_timestamp, timestamp):
        self.non_timestamp = non_timestamp
        self.timestamp = timestamp

test = MyClass(non_timestamp="Hello", timestamp=datetime.datetime.utcnow())

print(test)  # <__main__.MyClass object at 0x7f14e34aeca0>
print(test.non_timestamp)  # Hello
print(test.timestamp)  # 2023-06-15 06:12:29.930853

proxy = wrap_with_timestamp_localiser(test, tz_info=zoneinfo.ZoneInfo("America/Los_Angeles"))
# In reality its a TimestampLocaliserProxy
print(proxy)  # <__main__.TimestampLocaliserProxy object at 0x7f14e34aeb20>
# However type checkers think its not a proxy class
reveal_type(proxy) # Type of "proxy" is "MyClass"
print(proxy.non_timestamp)  # Hello
print(proxy.timestamp)  # 2023-06-14 13:12:29.930853-07:00
```
